src/ir/principal-type.cpp

```diff
@@ -0,0 +1,1062 @@
+  apply(assignments, type.exact);
+}
+
+void apply(const VarAssignments& assignments, VarHeapType& type) {
+  if (auto* i = std::get_if<Index>(&type)) {
+    type = assignments.heapTypes.get(*i);
+  }
+  if (auto* t = std::get_if<VarAbsHeapType>(&type)) {
+    apply(assignments, *t);
+  } else if (auto* t = std::get_if<VarDefHeapType>(&type)) {
+    apply(assignments, *t);
+  }
+}
+
+void apply(const VarAssignments& assignments, VarRef& ref) {
+  apply(assignments, ref.null);
+  apply(assignments, ref.ht);
+}
+
+void apply(const VarAssignments& assignments, VarType& type) {
+  if (auto* i = std::get_if<Index>(&type)) {
+    type = assignments.types.get(*i);
+  }
+  if (auto* t = std::get_if<VarRef>(&type)) {
+    apply(assignments, *t);
+    type = canonicalizeRef(*t);
+  }
+}
+
+void apply(const VarAssignments& assignments, PrincipalType& type) {
+  for (auto& t : type.rparams) {
+    apply(assignments, t);
+  }
+  for (auto& t : type.results) {
+    apply(assignments, t);
+  }
+}
+
+// matchBottom: join the bottom value into the assignemnt in `assignments` for
+// any variables appearing in the second parameter. Returns `true` if this
+// succeeds. TODO: This currently always succeeded, but if we supported repeated
+// variables in outputs properly, we would need to avoid mixing left assignment
+// and right assignemnt, so this would be able to fail.
+
+bool matchBottom(VarAssignments& assignments, const VarNullability& null) {
+  if (auto* i = std::get_if<Index>(&null)) {
+    return assignments.nulls.assign(*i, NonNullable);
+  }
+  return true;
+}
+
+bool matchBottom(VarAssignments& assignments, const VarSharedness& share) {
+  if (auto* i = std::get_if<Index>(&share)) {
+    return assignments.shares.assign(*i, BottomShare{});
+  }
+  return true;
+}
+
+bool matchBottom(VarAssignments& assignments, const VarExactness& exact) {
+  if (auto* i = std::get_if<Index>(&exact)) {
+    return assignments.exacts.assign(*i, Exact);
+  }
+  return true;
+}
+
+bool matchBottom(VarAssignments& assignments, const VarAbsHeapType& type) {
+  return matchBottom(assignments, type.share);
+}
+
+bool matchBottom(VarAssignments& assignments, const VarDefHeapType& type) {
+  return matchBottom(assignments, type.exact);
+}
+
+bool matchBottom(VarAssignments& assignments, const VarHeapType& type) {
+  if (auto* i = std::get_if<Index>(&type)) {
+    return assignments.heapTypes.assign(*i, BottomHeapType{});
+  } else if (auto* t = std::get_if<VarAbsHeapType>(&type)) {
+    return matchBottom(assignments, *t);
+  } else if (auto* t = std::get_if<VarDefHeapType>(&type)) {
+    return matchBottom(assignments, *t);
+  } else {
+    assert(std::get_if<BottomHeapType>(&type));
+    return true;
+  }
+}
+
+bool matchBottom(VarAssignments& assignments, const VarRef& ref) {
+  return matchBottom(assignments, ref.null) && matchBottom(assignments, ref.ht);
+}
+
+// match: Record the variable assignemnts necessary to make `a` match `b` (i.e.
+// to ensure `a` <: `b`).
```

Comment:
Continuing the above comment about the word "match", can we say `ensureSub`, as that IIUC what the comment describes?

Reply:
I think there's value in matching the spec wording, especially for something that is so directly taken from the spec, and also when the spec wording so conveniently pithy.

Reply:
Fair enough, sgtm.